Groupby hashmap fallback without sampling

Benkex opened the pull request.

Commit:
GroupBy optimization (switching to sorting if too many groups) 1.0

Commit:
GroupBy: Fix changes

Benkex committed:
GroupBy: Fix "merge results after switch to sort"

And implemented a lazy sampling

Benkex committed:
GroupBy: Fix "no aggregates" case

Benkex committed:
GroupBy: Final changes before PR

Benkex committed:
GroupBy: Revert changes in HashMapBenchmark

Benkex committed:
GroupBy: Revert changes 2.0

Benkex committed:
GroupBy: Fixes

1. Formatting
2. Test passes: GroupByOptimizations
.computeGroupByForSingleIndexScan
3. LOG(DEBUG) --> AD_LOG_DEBUG
ToDo: 21 more test cases, code logic change, ...

Benkex committed:
GroupBy: Fix test cases and code logic

Benkex committed:
GroupBy: Removed troublesome code

Now all tests run

Benkex committed:
GroupBy: Fix formatting

Benkex committed:
GroupBy: Modifications

Benkex committed:
GroupBy SampleFM: Start commit

Sample = Throw some entries into a HashMap.
If the no. of groups exceeds a threshold, we don't use HashMap.
FM = fully materialized

Benkex committed:
GroupBy Sample: Add test and benchmark, refactor

Benkex committed:
GroupBy Sample: Use Chao1 estimator, extend tests, format

Benkex committed:
GroupBy Sample: Refactor, remove benchmark

Replace debug print with a runtimeInfo Detail

Benkex committed:
GroupBy Sample: Remove code for other PR

Benkex committed:
GroupBy Sample: Fix, rem. unnecessary lib

Benkex committed:
GroupBy Sample: New class, refactor

Benkex committed:
GroupBy Sample: Format

Benkex committed:
GroupBy Sampe: Modifications

Benkex committed:
GroupBy Sample: Remove benchmark

Memory limit and other benchmark related changes
will be in another branch

Benkex committed:
GroupBy Sample: Remove obsolete imports

Benkex committed:
GroupBy Sample: Fixed tests

Benkex committed:
GroupBy Sample: Separate functionality, add timing

Benkex committed:
GBy HashMap fb: init

GBy = GroupBy, FB = fallback

Benkex committed:
GBy HashMap fb: Fix sample, extract code

Benkex committed:
GBy HashMap fb: Format

Benkex committed:
GBy HashMap fb: Separate helpers from test

Benkex committed:
GBy HashMap fb: Add tests

Benkex committed:
GBy HashMap fb: Add more tests

Couldn't fix the MultiColumn test - it fails

Benkex committed:
GBy HashMap fb: Add explanation

Benkex committed:
GBy HashMap fb: Fixed MultiColumn test

...by making helpers more generic

Benkex committed:
GBy HashMap fb: Fix include

Benkex committed:
GBy HashMap fb: Refactored helpers

Now we can test with lazy chunks

Benkex committed:
GBy HashMap fb: Add tests for lazy case

Benkex committed:
GBy HashMap fb: Turned functions into lambdas

This way we don't have to pass so many parameters

Benkex committed:
GBy HashMap fb: First round of modifications

Benkex committed:
GBy HashMap fb w/o sample: init

Benkex committed:
fb w/o sample: Fixes, don't copy IdTable

Instead we collect the indices of entries that either do or don't belong to the hashmap.
Then we update the hashmap with the collected table indices.
We repeat this for every block of entries, so this is a "semi-batch" approach.

Benkex committed:
fb w/o sample: Fix

I tried to export the creation of "ArrayOrVector<Id>" keys to a function but I failed. The function required a lambda, because the usecases are too different, it was ugly. Until I find a better solution, it may look like code duplication (though it may actually not be that way)

Benkex committed:
fb w/o sample: Small fixes

Benkex committed:
fb w/o sample: Refactor

joka921 reviewed (changes requested):
Thank you very much, 
This now looks much cleaner and understandable, let's fine-tune the details and benchmark this!

joka921 committed:
Fix the bug for the columns.

Signed-off-by: Johannes Kalmbach <johannes.kalmbach@gmail.com>

Benkex committed:
fb w/o sample: Separate functions, refactor

Benkex committed:
fb w/o sample: Extract functions, refactor

Benkex committed:
fb w/o sample: Use ValuesForTesting

Benkex committed:
fb w/o sample: Add more tests, Fixes

Altered makeProcessGroupsVisitor (we have to skip nonmatching entries there when switching to the hybrid fallback).
We skip the current block that we have just processed.
For "inWidth" we use _subtree->getResultWidth() instead of "beginIdTable.numColumns()" because it caused "bad optional access" errors.
Removed argument annotations. The IDE can display them.

Benkex committed:
fb w/o sample: Add more tests, Sort restTable

Added Aggregation tests to GroupByFallbackTest
Added tests for new functions to GroupByHashMapOptimizationTest
Fixed issue in GroupByImpl: We also had to sort restTable before doing doGroupBy

Benkex committed:
fb w/o sample: Reduce code

joka921 reviewed (changes requested):
I have looked at anything but the two large test files.
This already looks so much cleaner.

Benkex commented on src/engine/GroupByImpl.h:
```diff
@@ -319,14 +328,32 @@ class GroupByImpl : public Operation {
+
+    // Default ctor: no columns or aggregates set
+    HashMapOptimizationData() = default;
+    // Build with just aliases
+    explicit HashMapOptimizationData(
+        std::vector<HashMapAliasInformation> aliases)
+        : aggregateAliases_(std::move(aliases)) {}
+    // Build with aliases, columns, and aggregates
+    HashMapOptimizationData(std::vector<HashMapAliasInformation> aliases,
+                            std::vector<size_t> cols,
+                            std::vector<Aggregate> aggs)
```
Exactly

Benkex replied:
Yes, exactly

Benkex committed:
fb w/o sample: Implement sugestions

Implemented suggestions from this thread:
https://github.com/ad-freiburg/qlever/pull/2301#pullrequestreview-3196829848
TODO: Add tests for new helper functions in IdTableHelpers

Benkex committed:
fb w/o sample: Refining

Check sort instead of sorting in makeProcessGroupsVisitor
Rewrote comments
Replaced final sort in handleRemainderUsingHybridApproach to merging the two already-sorted tables
Created a test for the merging function

Benkex committed:
fb w/o sample: Test fix attempt & adjustments

Added explicit template instantiations at the end of GroupByImpl.cpp
Smaller adjustments to clean up the PR

Benkex committed:
fb w/o sample: Added test

Added test for processAggregateAliasesForBlock in GroupByHashMapOptimizationTest

joka921 commented on test/util/IdTableHelpers.h:
```diff
@@ -57,6 +57,47 @@ class CopyableIdTable : public TableImpl<N> {
 using IntOrId = std::variant<int64_t, Id>;
 using VectorTable = std::vector<std::vector<IntOrId>>;
 
+// Helper: construct a VectorTable from nested initializer lists of integers.
+static inline VectorTable makeVectorTable(
+    std::initializer_list<std::initializer_list<size_t>> rows) {
+  VectorTable vt;
+  for (auto rowIl : rows) {
+    VectorTable::value_type row;
+    for (auto v : rowIl) {
+      row.push_back(IntOrId(static_cast<int64_t>(v)));
+    }
+    vt.push_back(std::move(row));
+  }
+  return vt;
+}
```
I am not sure whether this function is necessary, can't you just do `VectorTable{{3, 4}, {7, 8}, {103, 102}};` directly, which is equivalent to this function?

joka921 commented on test/util/IdTableHelpers.h:
```diff
@@ -57,6 +57,47 @@ class CopyableIdTable : public TableImpl<N> {
+  return makeVectorTable(rows);
+}
+
+// Convenience helper: build an `IdTable` directly from nested integer
+// initializer lists, optionally supplying a transformation for plain ints.
+template <typename Transformation = decltype(ad_utility::testing::IntId)>
+IdTable makeIdTableFromInts(
+    std::initializer_list<std::initializer_list<size_t>> rows,
+    Transformation transformation = {}) {
+  return makeIdTableFromVector(makeVectorTable(rows),
+                               std::move(transformation));
+}
```
see above, I am not sure whether this is really necessary, or can be just achived by the existing constructors and helper functions, please check.

joka921 reviewed (changes requested):
I have left some comments, this code is complex, but we are getting there.

Benkex committed:
fb w/o sample: Implement suggestions

Reverted to sort instead of merging two already sorted tables
Code style changes
CPP_template usage for C++17 compatibility
Format fix
Attempt to fix automatic tests by adding explicit instantiation of updateHashMapWithTable to GroupByImpl.cpp

Benkex committed:
fb w/o sample: Benchmark, refactor

Added HybridGroupByBenchmark to measure time of te hybrid fallback strategy
Added detailed time measurements to GroupByImpl in the functions related to the hybrid fallback strategy
Added linear merge back, because it turned out to be much faster than sorting at the end of the hybrid fallback
Refactored setupGroupBy functions to be more usable

Benkex committed:
fb w/o sample: Sort-vs-merge Evaluation

When running HybridGroupByBenchmark, now it turned out that the difference between the two end-merging strategies is insignificant
In the next commit I am removing the merge code

Benkex committed:
fb w/o sample: Remove linear merge, rename

Removed linear merge algorithm and test from GroupByImpl which is not significantly faster than an Engine::sort in Release mode
Renamed "chunk" to "block" for conform usage of terminology

Benkex committed:
fb w/o sample: Mid-Table fallback

It enables exiting the hash map strategy when there is a large table which gets broken down into blocks in updateHashMapWithTable.
We now offset nonmatching indexes to enable tracing back the original table's entries.
Also this is probably the correct way to handle nonmatching indexes anyway

Set group threshold to 100 000 after evaluating with HybridGroupByBenchmark

Benkex committed:
fb w/o sample: Fix offset

We get SegFaults sometimes when  running HybridGroupByBenchmark.
Using "if (restTable.numRows() > 0)" helps but not always
When trying to return hybrid result we get this error:

GROUP_BY 0, N16sparqlExpression6detail15CountExpressionE0(#column_1#) AS 1
SORT(internal) on columns:asc(0)
Values for testing with 2 columns and 1000000 rows. -2936449197288137592 Supports limit: 0
terminate called after throwing an instance of 'ad_utility::AbortException'
  what():  Assertion `performCheck( varColMap, std::get<IdTableSharedLocalVocabPair>(data_).idTable_)` failed. Please report this to the developers. In file "/local/data-ssd/hargitab/qlever/src/engine/Result.cpp " at line 211
Aborted (core dumped)

Benkex committed:
fb w/o sample: Fix SegFault, refactor

Benkex committed:
fb w/o sample: Small fix

Benkex committed:
fb w/o sample: format and small fixes

Benkex committed:
fb w/o sample: Extend Benchmark

Benkex committed:
fb w/o sample: Refactor Benchmark

Benkex committed:
fb w/o sample: Fix Benchmark

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1668,117 +1540,335 @@ IdTable GroupByImpl::createResultFromHashMap(
+    // Merge the local vocab of each input block.
+    vocabMergeTimer.cont();
+    localVocab.mergeWith(inputLocalVocab);
+    vocabMergeTimer.stop();
```
clean up the timers. The vocabMergeTimer is probably not that interesting.

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1668,117 +1540,335 @@ IdTable GroupByImpl::createResultFromHashMap(
+  runtimeInfo().addDetail("hybridFinalMerge", finalMergeTimer.msecs());
+  runtimeInfo().addDetail("hybridFinalSort", finalSortTimer.msecs());
+  runtimeInfo().addDetail("hybridRowsProcessed", totalRowsProcessed);
+  runtimeInfo().addDetail("hybridNonMatchingRows", totalNonMatchingRows);
```
some of those are not that important.

joka921 reviewed (changes requested):
The first and most important step is to merge in the current `master`, s.t. we can test this actually. And Then I have some suggestions from this + a previous review, that you should incorporate (mostly cleaning up).

joka921 reviewed (commented):
Two additional suggestions.

Benkex committed:
Merging from upstream main

Benkex committed:
fb w/o sample: Complete merge

Applied runtimeparameter changes

There is an error with the boost container that I fixed with removing container from the CMakeLists.txt
But I didn't commit it, maybe it is just my local problem

Benkex committed:
Merge branch 'ad-freiburg:master' into groupby_hashmap_fallback_without_sampling

Benkex committed:
fb w/o sample: Apply suggestions

(in order of changes in GroupByImpl.cpp)
1. renamed onlyInsertPreexistingKeys to onlyUsePreexistingGroups
2. removed nonMatchingRowsOffset, replaced with ql::ranges::transform in updateHashMapWithTable
3. Moved RowToGroup struct into GroupLookupResult
4. Removed unimportant timers and counters from runtimeInfo
5. Added TODO to merge at the end of the hybrid fallback instead of sorting

(In GroupByImpl.h)
6. Added a description to RowToGroup
7. Refined the description of getHashEntries

Fixed import issue in GroupByStrategyHelpers.h

Benkex committed:
Merge branch 'ad-freiburg:master' into groupby_hashmap_fallback_without_sampling

joka921 reviewed (changes requested):
A thorough pass on anything but
1. The IdTableHelpers (those are in a separate PR and already approved)
2. The unit tests (I'll look at them once we have the results from the CI checks)

Overall this now looks very clean.

Benkex committed:
fb w/o sample: Apply suggestions 1

GroupByImpl.cpp:
1. Fixed efficiency problem: Not looking up the key twice in getHashEntries
2. Made `makeProcessGroupsVisitor` a class method
3. Added all template types to the `handleRemainderUsingHybridApproach` function call (code coverage fix)
4. ql::concepts::sentinel_for instead of std::sentinel_for
5. Removed unnecessary template from UpdateResult and explicit instantiation at the end

GroupByImpl.h:
7. Replaced default == operator definition by manual definition because of C++17

Benkex committed:
Merge branch 'master' into groupby_hashmap_fallback_without_sampling

Benkex committed:
fb w/o sample: Small fix

RowToGroup is now a member of GroupLookupResult

Benkex committed:
fb w/o sample: Skip non-matching values

Now we don't evaluate sub-expressions for the non-matching rows twice.
This is achieved with a modified generator `makeGeneratorSparse` that intelligently skips irrelevant values.
This approach is probably only better in C++17 for now.

Benkex committed:
fb w/o sample: Remove RowToGroup struct

It was unnecessary to map rows to groups, the two vectors are handled separately later.

Benkex committed:
fb w/o sample: Removed comment

Benkex committed:
Merge branch 'ad-freiburg:master' into groupby_hashmap_fallback_without_sampling

joka921 commented on src/engine/sparqlExpressions/SparqlExpressionGenerators.h:
```diff
@@ -200,6 +202,166 @@ CPP_template(typename Input, typename Transformation = ql::identity)(
+/// Return a generator that yields values only for the rows whose indices are
+/// contained in `indices`. The indices must be sorted in ascending order and
+/// must be strictly smaller than `numItems`.
+///
```
I haven't looked at this in detail, 
but it 
1. looks reasonable in principle, but very complicated (not sure if it can be done simpler) or already exists
2. It only solves the problem halfway (if there is a subexpression that is complicated, that still will be evaluated).

I think for now, you can leave your old solution, and if desired then keep this for a follow-up PR if required.

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1458,7 +1471,7 @@ GroupByImpl::HashMapAggregationData<NUM_GROUP_COLUMNS>::getHashEntries(
   }
 
-  return GroupLookupResult{std::move(matchedRowsToGroups),
+  return GroupLookupResult{std::move(matchedRows), std::move(groupIndexes),
```
Any particular reason why this has changed (can't you still couple the matched and the groupIndexes?

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1698,49 +1711,47 @@ IdTable GroupByImpl::createResultFromHashMap(
+      requires sparqlExpression::SingleExpressionResult<T> &&
+      VectorOfAggregationData<A>) {
+    const auto& matchedRows = groupLookupResult.matchedRows_;
+    const auto& groupIndexes = groupLookupResult.groupIndexes_;
```
Ah, you need the matched rows separately for your new sparse generator thing. you can keep it for now, as the more efficient fix for this also requires to have them separately probably.

joka921 commented on src/engine/GroupByImpl.cpp:
```diff
@@ -1698,49 +1711,47 @@ IdTable GroupByImpl::createResultFromHashMap(
+
+    auto generator = sparqlExpression::detail::makeGeneratorSparse(
+        std::forward<T>(singleResult), blockSize, evaluationContext,
+        std::move(indices));
```
1. no need to do a copy of the indices here. your `makeGeneratorSparse` should take some view thing (let us discuss this in person).

joka921 reviewed (changes requested):
Thank you very much, most of the stuff is  cleaned up now.

Benkex committed:
fb w/o sample: Apply suggestions

Benkex committed:
fb w/o sample: Fix inclue error

joka921 commented on src/engine/sparqlExpressions/SparqlExpressionGenerators.h:
```diff
@@ -200,6 +202,166 @@ CPP_template(typename Input, typename Transformation = ql::identity)(
+      while (currentIndex_ < target && genIt_ != genEnd_) {
+        ++genIt_;
+        ++currentIndex_;
+      }
```
```
ql::ranges::advance(genIt_, target - currentIndex_);
```

joka921 commented on src/engine/sparqlExpressions/SparqlExpressionGenerators.h:
```diff
@@ -200,6 +202,166 @@ CPP_template(typename Input, typename Transformation = ql::identity)(
+      while (currentIndex_ < nextTarget && genIt_ != genEnd_) {
+        ++genIt_;
+        ++currentIndex_;
+      }
```
```
++idxIt_;
skipToCurrentTarget();
```

joka921 commented on src/engine/sparqlExpressions/SparqlExpressionGenerators.h:
```diff
@@ -200,6 +202,166 @@ CPP_template(typename Input, typename Transformation = ql::identity)(
+    size_t currentIndex_ = 0;
+    size_t totalSize_ = 0;
+
+    void skipToCurrentTarget() {
```
make the previous target and current target parameters here, then this gets much simpler.

joka921 commented on src/engine/sparqlExpressions/SparqlExpressionGenerators.h:
```diff
@@ -200,6 +202,166 @@ CPP_template(typename Input, typename Transformation = ql::identity)(
+
+  iterator begin() { return iterator{&generator_, &indices_, totalSize_}; }
+
+  std::default_sentinel_t end() { return {}; }
```
I think you need `ql::default_sentinel_t` (not sure though)

joka921 reviewed (changes requested):
We can improve on the design on the sparse range thing (in particular move it to a separate file and add a unit test), the remainder looks rather clean now, we are almost there.

And pleeeasse fix the compilation (HybridBenchmark wrong include of ValuesForTesting, and maybe `ql::default_sentinel` for C++17 mode.

Benkex committed:
fb w/o sample: Implement Sparse generator

Benkex committed:
fb w/o sample: Format and build fix

Benkex committed:
fb w/o sample: Fix conan build error

Benkex committed:
fb w/o sample: Fix conan build error 2

Benkex committed:
fb w/o sample: Sonarcloud fixes

joka921 reviewed (changes requested):
Thank you very much. For your sparse view I will do a final pass before merging it,
but maybe you can have a look at the `ColumnIndex` business (or leave it as it was). I know that currently there is a mess because ColumnIndex is the same as size_t on linux, but not on Mac, which leads to trouble.

Also: try to address (when possible) sonarcloud issues in the code, but not in the benchmark, I will disable sonarcloud for the benchmark as I there honestly don't care too much:)

Benkex committed:
Create separate file for filterbyindicesview

Benkex committed:
Consistently use vector of ColumnIndex

Benkex committed:
Sonarcloud changes

Benkex committed:
Fix build error and add test

Benkex committed:
Minimal change to re-run build checks
